File: backend/src/resume_orchestrator/clients/llm_draft.py
from __future__ import annotations

import re
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Literal
from urllib.parse import quote

import httpx

from ..util.env import ResumeEnv

DraftProvider = Literal["openai", "anthropic", "google"]

_FENCED_LATEX_RE = re.compile(r"```(?:latex|tex)?\s*([\s\S]*?)```", re.IGNORECASE)


@dataclass(slots=True)
class DraftRequest:
    system: str
    user: str


DraftClient = Callable[[DraftRequest], Awaitable[str]]


async def _post_json(url: str, headers: dict[str, str], body: dict[str, object]) -> dict[str, object]:
    async with httpx.AsyncClient(timeout=httpx.Timeout(120.0)) as client:
        res = await client.post(url, headers=headers, json=body)
        if res.status_code >= 400:
            raise RuntimeError(f"HTTP {res.status_code}: {res.text}")
        return res.json()


async def _call_openai(api_key: str, req: DraftRequest) -> str:
    data = await _post_json(
        "https://api.openai.com/v1/chat/completions",
        headers={
            "Content-Type": "application/json",
            "Authorization": f"Bearer {api_key}",
        },
        body={
            "model": "gpt-4o-mini",
            "messages": [
                {"role": "system", "content": req.system},
                {"role": "user", "content": req.user},
            ],
            "temperature": 0.2,
        },
    )
    choices = data.get("choices") if isinstance(data, dict) else None
    if not choices:
        raise RuntimeError("OpenAI response missing message content")
    message = choices[0].get("message") if isinstance(choices[0], dict) else None
    content = message.get("content") if isinstance(message, dict) else None
    if not isinstance(content, str):
        raise RuntimeError("OpenAI response missing message content")
    return content


async def _call_anthropic(api_key: str, req: DraftRequest) -> str:
    data = await _post_json(
        "https://api.anthropic.com/v1/messages",
        headers={
            "Content-Type": "application/json",
            "x-api-key": api_key,
            "anthropic-version": "2023-06-01",
        },
        body={
            "model": "claude-3-5-sonnet-latest",
            "max_tokens": 4096,
            "system": req.system,
            "messages": [{"role": "user", "content": req.user}],
        },
    )
    content_blocks = data.get("content") if isinstance(data, dict) else None
    if not isinstance(content_blocks, list):
        raise RuntimeError("Anthropic response missing content[].text")
    for block in content_blocks:
        if isinstance(block, dict) and block.get("type") == "text":
            text = block.get("text")
            if isinstance(text, str):
                return text
    raise RuntimeError("Anthropic response missing content[].text")


async def _call_google(api_key: str, req: DraftRequest) -> str:
    url = (
        "https://generativelanguage.googleapis.com/v1beta/models/"
        f"gemini-1.5-flash:generateContent?key={quote(api_key, safe='')}"
    )
    data = await _post_json(
        url,
        headers={"Content-Type": "application/json"},
        body={
            "systemInstruction": {"role": "system", "parts": [{"text": req.system}]},
            "contents": [{"role": "user", "parts": [{"text": req.user}]}],
            "generationConfig": {"temperature": 0.2},
        },
    )
    candidates = data.get("candidates") if isinstance(data, dict) else None
    if not candidates:
        raise RuntimeError("Google response missing candidates[].content.parts[].text")
    content = candidates[0].get("content") if isinstance(candidates[0], dict) else None
    parts = content.get("parts") if isinstance(content, dict) else None
    if not isinstance(parts, list):
        raise RuntimeError("Google response missing candidates[].content.parts[].text")
    text = "".join(p.get("text", "") for p in parts if isinstance(p, dict))
    if not text:
        raise RuntimeError("Google response missing candidates[].content.parts[].text")
    return text


@dataclass(slots=True)
class ResolvedDraftProvider:
    provider: DraftProvider
    client: DraftClient


def resolve_draft_provider(env: ResumeEnv) -> ResolvedDraftProvider:
    provider: DraftProvider = env.RESUME_DRAFT_PROVIDER or "openai"
    if provider == "openai":
        if not env.OPENAI_API_KEY:
            raise RuntimeError("RESUME_DRAFT_PROVIDER=openai requires OPENAI_API_KEY.")
        api_key = env.OPENAI_API_KEY
        return ResolvedDraftProvider(provider=provider, client=lambda req: _call_openai(api_key, req))
    if provider == "anthropic":
        if not env.ANTHROPIC_API_KEY:
            raise RuntimeError("RESUME_DRAFT_PROVIDER=anthropic requires ANTHROPIC_API_KEY.")
        api_key = env.ANTHROPIC_API_KEY
        return ResolvedDraftProvider(provider=provider, client=lambda req: _call_anthropic(api_key, req))
    if provider == "google":
        if not env.GOOGLE_GENERATIVE_AI_API_KEY:
            raise RuntimeError("RESUME_DRAFT_PROVIDER=google requires GOOGLE_GENERATIVE_AI_API_KEY.")
        api_key = env.GOOGLE_GENERATIVE_AI_API_KEY
        return ResolvedDraftProvider(provider=provider, client=lambda req: _call_google(api_key, req))
    raise RuntimeError(f"Unknown RESUME_DRAFT_PROVIDER: {provider}")


def extract_latex_document(raw: str) -> str:
    fenced = _FENCED_LATEX_RE.search(raw)
    if fenced:
        return fenced.group(1).strip() + "\n"
    return raw.strip() + "\n"

File: backend/src/clients/llmDraft.ts
import { ResumeEnv } from "../util/env.js";

export type DraftRequest = {
  system: string;
  user: string;
};

export type DraftClient = (req: DraftRequest) => Promise<string>;

async function callOpenAI(apiKey: string, req: DraftRequest): Promise<string> {
  const res = await fetch("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`
    },
    body: JSON.stringify({
      model: "gpt-4o-mini",
      messages: [
        { role: "system", content: req.system },
        { role: "user", content: req.user }
      ],
      temperature: 0.2
    })
  });
  if (!res.ok) {
    throw new Error(`OpenAI HTTP ${res.status}: ${await res.text().catch(() => "")}`);
  }
  const data = (await res.json()) as { choices?: Array<{ message?: { content?: string } }> };
  const content = data.choices?.[0]?.message?.content;
  if (typeof content !== "string") throw new Error("OpenAI response missing message content");
  return content;
}

async function callAnthropic(apiKey: string, req: DraftRequest): Promise<string> {
  const res = await fetch("https://api.anthropic.com/v1/messages", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "x-api-key": apiKey,
      "anthropic-version": "2023-06-01"
    },
    body: JSON.stringify({
      model: "claude-3-5-sonnet-latest",
      max_tokens: 4096,
      system: req.system,
      messages: [{ role: "user", content: req.user }]
    })
  });
  if (!res.ok) {
    throw new Error(`Anthropic HTTP ${res.status}: ${await res.text().catch(() => "")}`);
  }
  const data = (await res.json()) as { content?: Array<{ type?: string; text?: string }> };
  const text = data.content?.find((c) => c.type === "text")?.text;
  if (typeof text !== "string") throw new Error("Anthropic response missing content[].text");
  return text;
}

async function callGoogle(apiKey: string, req: DraftRequest): Promise<string> {
  const url = `https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=${encodeURIComponent(apiKey)}`;
  const res = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      systemInstruction: { role: "system", parts: [{ text: req.system }] },
      contents: [{ role: "user", parts: [{ text: req.user }] }],
      generationConfig: { temperature: 0.2 }
    })
  });
  if (!res.ok) {
    throw new Error(`Google HTTP ${res.status}: ${await res.text().catch(() => "")}`);
  }
  const data = (await res.json()) as {
    candidates?: Array<{ content?: { parts?: Array<{ text?: string }> } }>;
  };
  const text = data.candidates?.[0]?.content?.parts?.map((p) => p.text ?? "").join("");
  if (!text) throw new Error("Google response missing candidates[].content.parts[].text");
  return text;
}

export type ResolvedDraftProvider = { provider: "openai" | "anthropic" | "google"; client: DraftClient };

export function resolveDraftProvider(env: ResumeEnv): ResolvedDraftProvider {
  const provider = env.RESUME_DRAFT_PROVIDER ?? "openai";
  if (provider === "openai") {
    if (!env.OPENAI_API_KEY) {
      throw new Error("RESUME_DRAFT_PROVIDER=openai requires OPENAI_API_KEY.");
    }
    const apiKey = env.OPENAI_API_KEY;
    return { provider, client: (req) => callOpenAI(apiKey, req) };
  }
  if (provider === "anthropic") {
    if (!env.ANTHROPIC_API_KEY) {
      throw new Error("RESUME_DRAFT_PROVIDER=anthropic requires ANTHROPIC_API_KEY.");
    }
    const apiKey = env.ANTHROPIC_API_KEY;
    return { provider, client: (req) => callAnthropic(apiKey, req) };
  }
  if (provider === "google") {
    if (!env.GOOGLE_GENERATIVE_AI_API_KEY) {
      throw new Error("RESUME_DRAFT_PROVIDER=google requires GOOGLE_GENERATIVE_AI_API_KEY.");
    }
    const apiKey = env.GOOGLE_GENERATIVE_AI_API_KEY;
    return { provider, client: (req) => callGoogle(apiKey, req) };
  }
  throw new Error(`Unknown RESUME_DRAFT_PROVIDER: ${provider as string}`);
}

export function extractLatexDocument(raw: string): string {
  const fenced = /```(?:latex|tex)?\s*([\s\S]*?)```/i.exec(raw);
  if (fenced) return fenced[1].trim() + "\n";
  return raw.trim() + "\n";
}
